### PR TITLE
feat: add cache(redis) controller

### PR DESCRIPTION
**Pull request by chlins**

### CODE REVIEW

File: pkg/cluster/cache/factory.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package cache
+
+import (
+	"github.com/goharbor/harbor-operator/pkg/lcm"
+)
+
+// NewCacheController returns a cache controller.
+// TODO add options
+func NewCacheController() lcm.Controller {
```
Comment (Ghostbaby) [inline pkg/cluster/cache/factory.go:9]: this func need to define which args need to input.

File: pkg/cluster/cache/redis.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+var (
+	redisFailoversGVR = redisOp.SchemeGroupVersion.WithResource(redisOp.RFNamePlural)
+)
+
+// NewRedisController is constructor for redis controller.
+func NewRedisController() lcm.Controller {
```
Comment (wangcanfengxs) [inline pkg/cluster/cache/redis.go:25]: Set the args to create a new RedisController

File: pkg/cluster/cache/factory.go (hunk 1 above)
Comment (chlins) [reply pkg/cluster/cache/factory.go:9]: Yup, this need input k8s.GetOptions, which defined in another pr, after the common k8s.GetOptions merged, i will add the option here.

File: pkg/cluster/cache/redis.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	ResourceManager    ResourceManager
+	expectCR, actualCR *redisOp.RedisFailover
+}
+
+// Apply creates/updates/scales the resources, like kubernetes apply operation.
+func (rc *RedisController) Apply(cluster *v1alpha2.HarborCluster) (*lcm.CRStatus, error) {
+	rc.Client.WithContext(rc.Ctx)
+	rc.DClient.WithContext(rc.Ctx)
+
+	crdClient := rc.DClient.WithResource(redisFailoversGVR).WithNamespace(cluster.Namespace)
```
Comment (wangcanfengxs) [inline pkg/cluster/cache/redis.go:47]: There need to check whether `cluster.inClusterCache` defined. If not defined. check the external Redis health and return the `CRStatus`

File: pkg/cluster/cache/readiness.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+	properties := lcm.Properties{}
+	properties.Add(cachePropertiesKey, rc.generateRedisSpec())
+	return cacheReadyStatus(&properties), nil
+}
+
+func (rc *RedisController) generateRedisSpec() *v1alpha2.ExternalRedisSpec {
+	return &v1alpha2.ExternalRedisSpec{
+		RedisHostSpec: harbormetav1.RedisHostSpec{
+			Host: rc.ResourceManager.GetService().Name,
```
Comment (wangcanfengxs) [inline pkg/cluster/cache/readiness.go:62]: Why directly call a new method `GetServiceName()`? So needn't generate the whole Service again.

File: pkg/cluster/cache/readiness.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+			Port: 6379,
+		},
+		RedisCredentials: harbormetav1.RedisCredentials{
+			PasswordRef: rc.ResourceManager.GetSecret().Name,
```
Comment (wangcanfengxs) [inline pkg/cluster/cache/readiness.go:66]: Same comment as above

**Review by wangcanfengxs — COMMENTED** [review]
comments left

File: pkg/cluster/cache/deploy.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+// It does not:
+// - perform any RedisFailovers downscale (left for downscale phase)
+// - perform any RedisFailovers upscale (left for upscale phase)
+// - perform any pod upgrade (left for rolling upgrade phase)
+func (rc *RedisController) Deploy(cluster *v1alpha2.HarborCluster) (*lcm.CRStatus, error) {
+	crdClient := rc.DClient.WithResource(redisFailoversGVR).WithNamespace(cluster.Namespace)
+	expectCR := rc.ResourceManager.GetCacheCR()
+
+	var err error
+	if err = controllerutil.SetControllerReference(cluster, expectCR.(metav1.Object), rc.Scheme); err != nil {
```
Comment (soulseen) [inline pkg/cluster/cache/deploy.go:29]: Which step is used to check if Redis exists?

**Comment by wangcanfengxs** [comment]
lgtm

**Review by Ghostbaby — APPROVED** [review]
lgtm

**Review by soulseen — APPROVED** [review]
/lgtm

**Pull request merged**